File: web/ray-tracer-ts/vec3.ts
export class Vec3 {
  constructor(
    public x: number,
    public y: number,
    public z: number,
  ) {}

  // Vector addition
  add(other: Vec3): Vec3 {
    return new Vec3(this.x + other.x, this.y + other.y, this.z + other.z);
  }

  // Vector subtraction
  sub(other: Vec3): Vec3 {
    return new Vec3(this.x - other.x, this.y - other.y, this.z - other.z);
  }

  // Scalar multiplication
  mul(scalar: number): Vec3 {
    return new Vec3(this.x * scalar, this.y * scalar, this.z * scalar);
  }

  // Scalar division
  div(scalar: number): Vec3 {
    return this.mul(1 / scalar);
  }

  // Component-wise multiplication (Hadamard product)
  multiply(other: Vec3): Vec3 {
    return new Vec3(this.x * other.x, this.y * other.y, this.z * other.z);
  }

  // Dot product
  dot(other: Vec3): number {
    return this.x * other.x + this.y * other.y + this.z * other.z;
  }

  // Length squared
  lengthSquared(): number {
    return this.dot(this);
  }

  // Length
  length(): number {
    return Math.sqrt(this.lengthSquared());
  }

  // Normalize (unit vector)
  normalize(): Vec3 {
    return this.div(this.length());
  }

  // Reflect vector around a normal
  reflect(normal: Vec3): Vec3 {
    // v - 2 * dot(v, n) * n
    return this.sub(normal.mul(2 * this.dot(normal)));
  }

  // Linear interpolation
  lerp(other: Vec3, t: number): Vec3 {
    return this.mul(1 - t).add(other.mul(t));
  }

  // Generate random vector in unit sphere
  static randomInUnitSphere(): Vec3 {
    while (true) {
      const p = new Vec3(
        -1 + 2 * Math.random(),
        -1 + 2 * Math.random(),
        -1 + 2 * Math.random(),
      );
      if (p.lengthSquared() < 1) {
        return p;
      }
    }
  }

  toString(): string {
    return `Vec3(${this.x}, ${this.y}, ${this.z})`;
  }
}

File: src/raytracer_raw/Vec3.scala
package raytracer_raw

import scala.scalajs.js
import scala.util.Random

// Direct port of TypeScript Vec3 class
// Mirrors web/ray-tracer-ts/vec3.ts
case class Vec3(x: Double, y: Double, z: Double) {

  // Vector addition
  def add(other: Vec3): Vec3 =
    Vec3(this.x + other.x, this.y + other.y, this.z + other.z)

  // Vector subtraction
  def sub(other: Vec3): Vec3 =
    Vec3(this.x - other.x, this.y - other.y, this.z - other.z)

  // Scalar multiplication
  def mul(scalar: Double): Vec3 =
    Vec3(this.x * scalar, this.y * scalar, this.z * scalar)

  // Scalar division
  def div(scalar: Double): Vec3 =
    this.mul(1.0 / scalar)

  // Component-wise multiplication (Hadamard product)
  def multiply(other: Vec3): Vec3 =
    Vec3(this.x * other.x, this.y * other.y, this.z * other.z)

  // Dot product
  def dot(other: Vec3): Double =
    this.x * other.x + this.y * other.y + this.z * other.z

  // Length squared
  def lengthSquared(): Double =
    this.dot(this)

  // Length
  def length(): Double =
    Math.sqrt(this.lengthSquared())

  // Normalize (unit vector)
  def normalize(): Vec3 =
    this.div(this.length())

  // Reflect vector around a normal
  def reflect(normal: Vec3): Vec3 = {
    // v - 2 * dot(v, n) * n
    this.sub(normal.mul(2.0 * this.dot(normal)))
  }

  // Linear interpolation
  def lerp(other: Vec3, t: Double): Vec3 =
    this.mul(1.0 - t).add(other.mul(t))

  override def toString: String =
    s"Vec3($x, $y, $z)"
}

object Vec3 {
  // Generate random vector in unit sphere
  def randomInUnitSphere(): Vec3 = {
    var p = Vec3(0, 0, 0)
    var continue = true
    while (continue) {
      p = Vec3(
        -1.0 + 2.0 * Math.random(),
        -1.0 + 2.0 * Math.random(),
        -1.0 + 2.0 * Math.random()
      )
      if (p.lengthSquared() < 1.0) {
        continue = false
      }
    }
    p
  }
}
